app/scripts/comp/launcher-cordova.js

```diff
@@ -0,0 +1,216 @@
+        const basePath = this.getDataPath().join('/');
+        const createDir = (dirEntry, path, callback) => {
+            const name = path.shift();
+            dirEntry.getDirectory(name, { create: true }, dirEntry => {
+                if (path.length) { // there is more to create
+                    createDir(dirEntry, path, callback);
+                } else {
+                    callback();
+                }
+            }, callback);
+        };
+
+        const localPath = dir.replace(basePath, '').split('/').filter(s => !!s);
+
+        if (localPath.length) {
+            window.resolveLocalFileSystemURL(basePath, dirEntry => {
+                createDir(dirEntry, localPath, callback);
+            }, callback);
+        } else {
+            callback();
+        }
+    },
+    parsePath: function(fileName) {
+        const parts = fileName.split('/');
+
+        return {
+            path: fileName,
+            dir: parts.pop(),
+            file: parts.join('/')
+        };
+    },
+    createFsWatcher: function(path) {
+        return null; // not in android with content provider
+    },
+    // ensureRunnable: function(path) { },
+    preventExit: function(e) {
+        e.returnValue = false;
+        return false;
+    },
+    exit: function() {
+        this.hideApp();
+    },
+    requestExit: function() { /* skip in cordova */ },
+    requestRestart: function() {
+        window.location.reload();
+    },
+    cancelRestart: function() { /* skip in cordova */ },
+    setClipboardText: function(text) {
+        // TODO
+    },
+    getClipboardText: function() {
+        // TODO
+    },
+    clearClipboardText: function() {
+        // TODO
+    },
+    minimizeApp: function() {
+        this.hideApp();
+    },
+    canMinimize: function() {
+        return false;
+    },
+    updaterEnabled: function() {
+        return false;
+    },
+    // getMainWindow: function() { },
+    resolveProxy: function(url, callback) { /* skip in cordova */ },
+    openWindow: function(opts) { /* skip in cordova */ },
+    hideApp: function() { /* skip in cordova */ },
+    isAppFocused: function() {
+        return false; /* skip in cordova */
+    },
+    showMainWindow: function() { /* skip in cordova */ },
+    // spawn: function(config) { },
+    openFileChooser: function(callback, button) {
+        const onFileSelected = function(selected) {
+            window.resolveLocalFileSystemURL(selected.uri, fileEntry => {
+                fileEntry.file(file => {
+                    file.path = file.localURL;
+                    file.name = selected.name;
+                    callback(null, file);
+                });
+            });
+        };
+
+        window.cordova.exec(onFileSelected, callback, 'FileChooser', 'choose');
+    },
+
+    fingerprints: {
+        config: {
+            disableBackup: true,
+            clientId: 'keeweb'
+        },
+
+        register: function(appModel, fileInfo, password) {
```

Comment:
Of course, I mean the conceptual return. I.e. for async method you should add a callback or promise.